compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
   int64_t targetVectorSize = 8;
-  if (llvm::any_of(distConfig.minTileSizes, [&](int64_t tileSize) {
+  if (llvm::any_of(minTileSizes, [&](int64_t tileSize) {
         return tileSize > 1 && tileSize < targetVectorSize;
       })) {
-    return failure();
+    return;
   }
 
-  // Target 16x16 tile sizes if there are AVX512 features and all the tile sizes
-  // are greater than or equal to 16.
+  // Target 16x16 tile sizes if there are AVX512 features and all the tile
+  // sizes are greater than or equal to 16.
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(entryPointFn);
   if (hasAVX512fFeature(targetAttr) &&
-      llvm::all_of(distConfig.minTileSizes, [](int64_t tileSize) {
+      llvm::all_of(minTileSizes, [](int64_t tileSize) {
         return tileSize == 1 || tileSize >= 16;
       })) {
     targetVectorSize = 16;
   }
 
   // Replace dims to be vectorized with the new tile sizes.
   std::replace_if(
-      distConfig.minTileSizes.begin(), distConfig.minTileSizes.end(),
+      minTileSizes.begin(), minTileSizes.end(),
       [](int64_t tileSize) { return tileSize > 1; }, targetVectorSize);
 
+  sizes = minTileSizes;
+  scalableSizeFlags = SmallVector<bool>(sizes.size(), false);
+}
+
+/// Utility to compute the tile sizes of transposes for AArch64 SME. Unlike
+/// other targets, the tile sizes picked here must exactly match multiples of
+/// the SME hardware virtual tiles, as there is currently no support for
+/// lowering non-standard shapes.
+static void
+getTransposeAArch64SMEVectorSizes(linalg::GenericOp genericOp,
+                                  SmallVectorImpl<int64_t> &sizes,
+                                  SmallVectorImpl<bool> &scalableSizeFlags) {
+  if (!transposeLoweringPreconditionAArch64SME(genericOp))
+    return;
+
+  auto elementType = nonWideningLinalgElementType(genericOp);
+  if (failed(elementType))
+    return;
+
+  if (elementType->isF32()) {
+    sizes.append({4, 4});
+    scalableSizeFlags.append({true, true});
+  }
+
+  if (elementType->isF64()) {
+    sizes.append({2, 2});
+    scalableSizeFlags.append({true, true});
+  }
+}
+
+/// Main utility to compute the transpose vector sizes.
+static std::optional<SizesAndScalableFlags>
+getTransposeVectorSizes(mlir::FunctionOpInterface entryPointFn,
+                        linalg::GenericOp genericOp,
+                        const LinalgOpInfo &linalgOpInfo,
+                        const TargetMLTransformInfo &targetMLTransInfo) {
+  SmallVector<int64_t> tileSizes;
+  SmallVector<bool> scalableFlags;
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(entryPointFn);
+  if (hasAVX2Feature(targetAttr))
+    getTransposeX86VectorSizes(entryPointFn, genericOp, linalgOpInfo,
+                               targetMLTransInfo, tileSizes, scalableFlags);
+  else if (isAArch64(targetAttr) && hasSMEFeature(targetAttr) &&
+           clEnableScalableVectorization)
+    getTransposeAArch64SMEVectorSizes(genericOp, tileSizes, scalableFlags);
```

Comment:
few nits:

1. It is better to use `isX86` for the first condition.
2. style nit: wrap `braces` for the if-else statements. This is the few difference between IREE and MLIR in terms of coding style.
3. Can we move the `isSMEFeature` and `clEnableScalableVectorization` logics to `getTransposeAArch64SMEVectorSizes`, and rename it to `getTransposeAArch64VectorSizes`? Today, most of the methods are separated based on the `isX86`, `isRISCV`, and `isAArch64`. IMO, keeping it this way is better for layering. I'm afraid that this if-statements becomes a monster and adds complexities.

Reply:
much nicer 👍